hazelcast/src/hazelcast/client/compact.cpp

```diff
@@ -0,0 +1,678 @@
+    }
+}
+void
+compact_writer::write_string(const std::string& field_name,
+                             const boost::optional<std::string>& value)
+{
+    if (default_compact_writer != nullptr) {
+        default_compact_writer->write_string(field_name, value);
+    } else {
+        schema_writer->add_field(field_name, pimpl::field_kind::STRING);
+    }
+}
+namespace pimpl {
+compact_reader
+create_compact_reader(
+  pimpl::compact_stream_serializer& compact_stream_serializer,
+  object_data_input& object_data_input,
+  const pimpl::schema& schema)
+{
+    return compact_reader{ compact_stream_serializer,
+                           object_data_input,
+                           schema };
+}
+} // namespace pimpl
+compact_reader::compact_reader(
+  pimpl::compact_stream_serializer& compact_stream_serializer,
+  serialization::object_data_input& object_data_input,
+  const pimpl::schema& schema)
+  : compact_stream_serializer(compact_stream_serializer)
+  , object_data_input(object_data_input)
+  , schema(schema)
+{
+    uint32_t final_position;
+    size_t number_of_var_size_fields = schema.number_of_var_size_fields();
+    if (number_of_var_size_fields != 0) {
+        uint32_t data_length = object_data_input.read<int32_t>();
+        data_start_position = object_data_input.position();
+        variable_offsets_position = data_start_position + data_length;
+        if (data_length < pimpl::offset_reader::BYTE_OFFSET_READER_RANGE) {
+            get_offset = pimpl::offset_reader::get_offset<int8_t>;
+            final_position =
+              variable_offsets_position + number_of_var_size_fields;
+        } else if (data_length <
+                   pimpl::offset_reader::SHORT_OFFSET_READER_RANGE) {
+            get_offset = pimpl::offset_reader::get_offset<int16_t>;
+            final_position =
+              variable_offsets_position +
+              (number_of_var_size_fields * util::Bits::SHORT_SIZE_IN_BYTES);
+        } else {
+            get_offset = pimpl::offset_reader::get_offset<int32_t>;
+            final_position =
+              variable_offsets_position +
+              (number_of_var_size_fields * util::Bits::INT_SIZE_IN_BYTES);
+        }
+    } else {
+        get_offset = pimpl::offset_reader::get_offset<int8_t>;
+        variable_offsets_position = 0;
+        data_start_position = object_data_input.position();
+        final_position =
+          data_start_position + schema.fixed_size_fields_length();
+    }
+    // set the position to final so that the next one to read something from
+    // `in` can start from correct position
+    object_data_input.position(final_position);
+}
+
+bool
+compact_reader::is_field_exists(const std::string& field_name,
+                                enum pimpl::field_kind kind) const
+{
+    const auto& fields = schema.fields();
+    const auto& field_descriptor = fields.find(field_name);
+    if (field_descriptor == fields.end()) {
+        return false;
+    }
```

Comment:
This retrieves the value and if it can not be found returns an iterator equal to `fields.end()` . If I refactor this to have a fieldExists method, I need to do two lookups to map.